### PR TITLE
HHH-11768 / HHH-11714 - Fix FK ConstraintViolationException Ordered Inserts

### PR DESCRIPTION
**Pull request by Naros**
https://hibernate.atlassian.net/browse/HHH-11768 - FK ConstraintViolation Cascade on Persist
https://hibernate.atlassian.net/browse/HHH-11714 - FK ConstraintViolation Secondary Table w/Inheritance

I've added both of these JIRA fixes to the same PR since the final fix for both was very similar with the second one requiring a little more code for inheritance-based classes.

### CODE REVIEW

**Comment by Naros** [comment]
The problem with HHH-11768 was that the ordering wasn't being handled correctly when the batches were arranged as follows:
<pre>
MarketBidResult: {
  parentEntities: [ MarketBid ], 
  childEntities: []
},
MarketBidGroup: {
  parentEntities: [],
  childEntities: [ MarketBid ]
},
MarketBid: [
  parentEntities: [ MarketBidGroup ],
  childEntities: []
}
</pre>
Insert ordering ended up yielding the following when inspecting `MarketBid` and doing the parent/child swaps.
<pre>
MarketBid
MarketBidResult
MarketBidGroup
</pre>
The key here was to add one additional condition in the second inner-loop for this case:
<pre>
else if ( batchIdentifier.getParentEntityNames().contains( nextBatchIdentifier.getEntityName() ) ) {
  latestBatches.remove( batchIdentifier );
  latestBatches.add( j, batchIdentifier );
}
</pre>
That forces the final sort to end up ordered correctly as:
<pre>
MarketBidGroup
MarketBid
MarketBidResult
</pre>

The fix for HHH-11714 builds off the above fix but requires that we also handle checking for the root entity name for the situations where a mapping is using the root entity type class name but the actual insert operation entity name is that of a subclass.

For example, when inserting the `ShapeCircleEntity`, it was being compared name-wise against `ShapeEntity` as that is what the mapping used in `GeographicArea`.  The problem was that forced `ShapeCircleEntity` to always be at the end of the sorted insertion list, which caused a FK constraint exception.

The non-sorted insert order is:
<pre>
TopLevel
Shape (Polygon)
GeographicArea (polygon)
Shape (Circle)
Shape (Circle secondary table)
GeographicArea (circle)
</pre>
When applying sorting we had this before the fix:
<pre>
TopLevel
Shape (Polygon)
GeographicArea (polygon)
GeographicArea (circle)
Shape (Circle)
Shape (Circle secondary table)
</pre>
The desired order after the fix is:
<pre>
TopLevel
Shape (Polygon)
Shape (Circle)
Shape (Circle secondary table)
GeographicArea (polygon)
GeographicArea (circle)
</pre>
Allowing the BatchIdentifier to maintain a root-entity-name allows us to check for that condition in the sort so make sure that subclasses are ordered properly.

File: hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java [hunk 1]
```diff
@@ -1049,13 +1050,35 @@ public String getEntityName() {
...
 			}
+
+			public boolean hasAnyParentEntityNames(String... entityNames) {
+				for ( String entityName : entityNames ) {
+					if ( parentEntityNames.contains( entityName ) ) {
+						return true;
+					}
+				}
+				return false;
+			}
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java]: what about 
```
public boolean hasAnyParentEntityNames(BatchIdentifier batchIdentifier) {
	if ( parentEntityNames.contains( batchIdentifier.getEntityName() )
		|| parentEntityNames.contains( batchIdentifier.getRootEntityName() ) ) {
		return true;
	}
	return false;
}
```

File: hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java [hunk 2]
```diff
@@ -1049,13 +1050,35 @@ public String getEntityName() {
...
+
+			public boolean hasAnyChildEntityNames(String... entityNames) {
+				for ( String entityName : entityNames ) {
+					if ( childEntityNames.contains( entityName ) ) {
+						return true;
+					}
+				}
+				return false;
+			}
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java]: and 
```
public boolean hasAnyChildEntityNames(BatchIdentifier batchIdentifier) {
	if ( childEntityNames.contains( batchIdentifier.getEntityName() )
		|| childEntityNames.contains( batchIdentifier.getRootEntityName() ) ) {
		return true;
	}
	return false;
}
```

**Comment by Naros** [comment]
Applied upstream

**Pull request closed without merging**